src/backend/optimizer/path/allpaths.c

```diff
@@ -2125,15 +2125,38 @@ set_cte_pathlist(PlannerInfo *root, RelOptInfo *rel, RangeTblEntry *rte)
 			subplan = subquery_planner(cteroot->glob, subquery, cteroot, cte->cterecursive,
 									   tuple_fraction, &subroot, config);
 
-			cteplaninfo->shared_plan = prepare_plan_for_sharing(cteroot, subplan);
+			if (subplan->flow->locustype != CdbLocusType_SegmentGeneral)
+			{
+				cteplaninfo->subplan = prepare_plan_for_sharing(cteroot, subplan);
+			}
+			else
+			{
+				cteplaninfo->subplan = subplan;
+			}
+
 			cteplaninfo->subroot = subroot;
 		}
 
 		/*
 		 * Create another ShareInputScan to reference the already-created
-		 * subplan.
+		 * subplan if not avoiding sharing. Sharing SegmentGeneral subplan may
+		 * lead to deadlock when executed with 1-gang and joined with n-gang.
 		 */
-		subplan = share_prepared_plan(cteroot, cteplaninfo->shared_plan);
+		if (cteplaninfo->subplan->flow->locustype != CdbLocusType_SegmentGeneral)
+		{
+			subplan = share_prepared_plan(cteroot, cteplaninfo->subplan);
+		}
+		else
```

Comment:
Can we simplify it like this?
```suggestion
		else if (subplan == NULL)
```